### PR TITLE
[java] Expose symbol annotations

### PR DESCRIPTION
**Pull request by jsotuyod**
## Describe the PR

This PR takes on @oowekyala's work on symbol annotations (see #4241) and expands on it to finalize it with a few minor improvements along the way.

Notice this PR focuses solely on symbolic annotations, that is, those that Java reflection will retrieve through `AnnotatedElement.getDeclaredAnnotations()`, which excludes both those with `Target` `ElementType.TYPE_USE` and `ElementType.TYPE_PARAMETER`, as those are actually impacting on the types (and in the reflection API obtainable only through `getAnnotated*Type()` family of methods. Targeting those are a next step, but both, independent, and massive enough to be worth a separate PR.

## Related issues

- Supercedes #4241 and #4240
- Partially covers #3752 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/GenericSigBase.java [hunk 1]
```diff
@@ -250,5 +255,16 @@ public JTypeMirror getReturnType() {
         public String toString() {
             return signature;
         }
+
+        void addParameterAnnotation(int paramIndex, SymAnnot annot) {
+            if (parameterAnnotations.isEmpty()) {
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/GenericSigBase.java:260]: One thing is the annotations you are adding here are the annotations on the variable, not the type annotations. Accordingly, they shouldn't be in the method signature, but stay in the symbol (ExecutableStub)

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/ast/AbstractAstExecSymbol.java [hunk 2]
```diff
@@ -35,6 +41,15 @@ protected AbstractAstExecSymbol(T node, AstSymFactory factory, JClassSymbol owne
...
         );
+        
+        NodeStream<ASTAnnotation> annotStream = node.getDeclaredAnnotations();
+        if (annotStream.isEmpty()) {
+            declaredAnnotations = Collections.emptyList();
+        } else {
+            final List<SymAnnot> annotations = new ArrayList<>();
+            annotStream.forEach(n -> annotations.add(new AstSymbolicAnnot(n)));
+            declaredAnnotations = Collections.unmodifiableList(annotations);
+        }
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/ast/AbstractAstExecSymbol.java:52]:
```suggestion
            declaredAnnotations = Collections.unmodifiableList(annotStream.toList(ASTSymbolicAnnot::new));
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/ast/AbstractAstVariableSym.java [hunk 3]
```diff
@@ -18,9 +25,20 @@
...
         super(node, factory);
+        
+        NodeStream<ASTAnnotation> annotStream = node.getDeclaredAnnotations();
+        if (annotStream.isEmpty()) {
+            declaredAnnotations = Collections.emptyList();
+        } else {
+            final List<SymAnnot> annotations = new ArrayList<>();
+            annotStream.forEach(n -> annotations.add(new AstSymbolicAnnot(n)));
+            declaredAnnotations = Collections.unmodifiableList(annotations);
+        }
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/ast/AbstractAstVariableSym.java:41]:
```suggestion
            declaredAnnotations = Collections.unmodifiableList(annotStream.toList(ASTSymbolicAnnot::new));
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/ast/AstClassSym.java [hunk 4]
```diff
@@ -116,6 +121,14 @@
...
             }
         }
+        
+        final NodeStream<ASTAnnotation> annotStream = node.getDeclaredAnnotations();
+        if (annotStream.nonEmpty()) {
+            myAnnotations = new ArrayList<>();
+            annotStream.forEach(anode -> myAnnotations.add(new AstSymbolicAnnot(anode)));
+        } else {
+            myAnnotations = Collections.emptyList();
+        }
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/ast/AstClassSym.java:131]:
```suggestion
            this.declaredAnnotations = Collections.unmodifiableList(annotStream.toList(ASTSymbolicAnnot::new));
```

The list is not touched anywhere else. So I don't think you need `myAnnotations`

**Review by oowekyala — COMMENTED** [review]
This looks fine to me, thanks! I have a couple of minor comments. 

I implemented *type* annotation recovery from ASM in [this branch](https://github.com/oowekyala/pmd/tree/clem.type-annotations-support). This does no interfere with your effort I think, but that may be a candidate for the next PR, to avoid diverging

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/GenericSigBase.java (hunk 1 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/GenericSigBase.java:260]: See 2df039f4872d1fd6c925be2e5e08da76ddee87c8 for the relevant changes

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 32 violations,<br />
introduces 28 new violations, 0 new errors and 0 new configuration errors,<br />
removes 20 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/d60c627a0dde409598545293377d9912/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 50508 violations,<br />
introduces 34670 new violations, 9 new errors and 0 new configuration errors,<br />
removes 138301 violations, 4 errors and 7 configuration errors.<br />
<a href="https://chunk.io/pmd/d60c627a0dde409598545293377d9912/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/GenericSigBase.java (hunk 1 above)
Comment (jsotuyod) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/GenericSigBase.java]: Great point, this should be fixed now.

**Comment by jsotuyod** [comment]
This should be ready now!

@oowekyala, as for how to let type annotations flow through type inference, keeping equality and what not… I've been thinking, maybe an option here is to do something similar to what Java'a reflection API actually does. Have the "annotated" types be wrappers on the standard types, and expose both (ie: `Method.getReturnType()` vs `Method.getAnnotatedReturnType()`). Equality can always be delegated to the not annotated one, and using a decorator can help minimize the number of classes we need to cope with.

I'm yet to take a look into what you did with https://github.com/oowekyala/pmd/tree/clem.type-annotations-support, but hope this helps. Let me know if I can assist you in any way with this.

**Pull request merged**